--- step-005/main.odin
+++ step-006/main.odin
@@ -24,7 +24,19 @@
 	token_index: int,
 }
 
-tokenize :: proc(source_text: string) -> [dynamic]Token {
+TokenizationErrorType :: enum {
+	InvalidNumber,
+}
+
+TokenizationError :: struct {
+	type: TokenizationErrorType,
+	start_byte: int,
+	line_start_byte: int,  // To find the line length, just iterate until next \n
+	line_number: int,
+	column_number: int,
+}
+
+tokenize :: proc(source_text: string) -> ([dynamic]Token, [dynamic]TokenizationError) {
 	// TODO: Assert that source file length is < INT_MAX
 	// TODO: Assert that source file ends with whitespace
 	when ODIN_DEBUG {
@@ -35,8 +47,10 @@
 	}
 
 	output := make([dynamic]Token, context.temp_allocator)
+	errors := make([dynamic]TokenizationError, context.temp_allocator)
 
 	line_number := 1
+	line_start_byte := 0
 	column_number := 0
 	token_index := 0
 
@@ -46,6 +60,7 @@
 	// we loop through characters.
 	currently_tokenizing := false
 	current_token : Token
+	skip_to_next_token_on_error := false
 	// Loop through each byte
 	for next_position < len(source_text) {
 		current_position = next_position
@@ -73,12 +88,16 @@
 				append(&output, current_token)
 				current_token = {}
 				currently_tokenizing = false
+			} else if skip_to_next_token_on_error {
+				// Reset now that we are going to next token
+				skip_to_next_token_on_error = false
 			}
 
 			// A newline increments line number and resets column number
 			if character == '\n' {
 				line_number += 1
 				column_number = 0
+				line_start_byte = current_position + 1
 			}
 
 			when ODIN_DEBUG {
@@ -87,6 +106,8 @@
 				}
 			}
 			continue
+		} else if skip_to_next_token_on_error {
+			continue
 		}
 
 		// Something other than whitespace! Let's figure out what it is
@@ -111,11 +132,46 @@
 
 			token_index += 1
 			currently_tokenizing = true
+			continue
+		}
+
+		// Check that the next character isn't obviously syntactically incorrect
+		if
+			current_token.type == .ConstantInteger &&
+			((character < '0' || character > '9') && character != '_')
+		{
+			error := TokenizationError {
+				type = .InvalidNumber,
+				start_byte = current_position,
+				line_start_byte = line_start_byte,
+				line_number = line_number,
+				column_number = column_number,
+			}
+			append(&errors, error)
+			skip_to_next_token_on_error = true
+			currently_tokenizing = false
+			current_token = {}
+			token_index -= 1
+
+			when ODIN_DEBUG {
+				end := line_start_byte;
+				for end < len(source_text) && source_text[end] != '\n'{
+					end += 1
+				}
+				log.debugf(
+					"> Token Error: %v (%v:%v, byte %v) (%v)",
+					source_text[error.line_start_byte:end],
+					error.line_number,
+					error.column_number,
+					error.start_byte,
+					error.type
+				)
+			}
 		}
 	}
 
 	// Assume source text ends in whitespace, so final token gets appended
-	return output
+	return output, errors
 }
 
 token_to_string :: proc(token: Token, source_text: string) -> string {
@@ -126,6 +182,10 @@
 	return character == ' ' || character == '\t' || character == '\n'
 }
 
+print_tokenization_errors :: proc() {
+
+}
+
 ////////////////////////////////////////////////////////////////////////////////
 // Tests
 //
@@ -136,7 +196,7 @@
 @(test)
 test_tokenize :: proc(t: ^testing.T) {
 	source_text := "This   is my program\nLine two\n Line  three \n"
-	output := tokenize(source_text)
+	output, _ := tokenize(source_text)
 	testing.expect_value(t, token_to_string(output[0], source_text), "This")
 	testing.expect_value(t, output[0].line_number, 1)
 	testing.expect_value(t, output[0].column_number, 1)
@@ -174,7 +234,7 @@
 @(test)
 test_tokenize_005 :: proc(t: ^testing.T) {
 	source_text := "my_var_1 := 1_337\nmy_var_2 := 663\n"
-	output := tokenize(source_text)
+	output, _ := tokenize(source_text)
 	testing.expect_value(t, token_to_string(output[0], source_text), "my_var_1")
 	testing.expect_value(t, output[0].line_number, 1)
 	testing.expect_value(t, output[0].column_number, 1)
@@ -200,3 +260,15 @@
 	testing.expect_value(t, output[5].column_number, 13)
 	testing.expect_value(t, output[5].token_index, 5)
 }
+
+@(test)
+test_tokenize_006 :: proc(t: ^testing.T) {
+	source_text := "1_my_var_ := 1_337\nmy_var_2 := 6^3\n"
+	_, errors := tokenize(source_text)
+	testing.expect_value(t, errors[0].type, TokenizationErrorType.InvalidNumber)
+	testing.expect_value(t, errors[0].line_number, 1)
+	testing.expect_value(t, errors[0].column_number, 3)
+	testing.expect_value(t, errors[1].type, TokenizationErrorType.InvalidNumber)
+	testing.expect_value(t, errors[1].line_number, 2)
+	testing.expect_value(t, errors[1].column_number, 14)
+}
