--- step-007/main.odin
+++ step-008/main.odin
@@ -5,6 +5,8 @@
 import "core:log"
 import vmem "core:mem/virtual"
 import "core:os"
+import "core:strings"
+import "core:unicode/utf8"
 
 when !ODIN_DEBUG { _ :: log } // Avoid unused import error for log
 
@@ -20,6 +22,8 @@
 
 	input_buffer: [1024]byte
 	for {
+		defer free_all(context.temp_allocator)
+
 		fmt.print("> ")
 		n, err := os.read(os.stdin, input_buffer[:])
 		if err != nil {
@@ -27,13 +31,15 @@
 			return
 		}
 
+		source_text := string(input_buffer[:n])
+
 		// Tokenize input
-		tokens, errors := tokenize(string(input_buffer[:n]), arena_allocator)
+		tokens, errors := tokenize(source_text, arena_allocator)
 
 		// Print out any errors
 		if len(errors) > 0 {
 			for error in errors {
-				fmt.eprintln("Error: ", error)
+				fmt.eprint(generate_error_message(error, source_text))
 			}
 			continue
 		}
@@ -216,6 +222,60 @@
 	column_number: int,
 }
 
+// Generate a nice-looking error message from a tokenization error.
+// Allocates using context's temporary allocator; free using `free_all(context.temp_allocator)`
+//
+// Example output:
+//   Error: Tokenization: Invalid Number (1:12; byte 11)
+//       a := 234234h //extra line context
+//       -----------^
+generate_error_message :: proc(error: Error, source_text: string) -> string {
+	// Use a string builder to generate a string with multiple lines
+	builder := strings.builder_make(context.temp_allocator)
+	strings.write_string(&builder, "Error: ")
+
+	// Print out the error information
+	switch (error.type) {
+	case .TokenizationInvalidNumber:
+		strings.write_string(&builder, "Tokenization: Invalid Number")
+	}
+	strings.write_string(&builder, " (")
+	strings.write_int(&builder, error.line_number)
+	strings.write_rune(&builder, ':')
+	strings.write_int(&builder, error.column_number)
+	strings.write_string(&builder, "; byte ")
+	strings.write_int(&builder, error.start_byte)
+	strings.write_string(&builder, ")\n")
+
+	// Get the byte location for the end of the line
+	line_end_byte := error.line_start_byte
+	for character in source_text[error.line_start_byte:] {
+		if character == '\n' {
+			break
+		} else {
+			line_end_byte += utf8.rune_size(character)
+		}
+		// TODO: Limit line length to a maximum?
+	}
+
+	// Print out the line in the source text
+	strings.write_string(&builder, "    ") // Indentation
+	strings.write_string(&builder, source_text[error.line_start_byte:line_end_byte])
+	strings.write_rune(&builder, '\n')
+
+	// Print an arrow to indicate which column in the source text line has the error
+	strings.write_string(&builder, "    ") // Indentation
+	cursor := 1
+	for cursor < error.column_number {
+		strings.write_rune(&builder, '-')
+		cursor += 1
+	}
+	strings.write_rune(&builder, '^')
+	strings.write_rune(&builder, '\n')
+
+	return strings.to_string(builder)
+}
+
 ////////////////////////////////////////////////////////////////////////////////
 // Tests
 ////////////////////////////////////////////////////////////////////////////////
