--- step-008/main.odin
+++ step-009/main.odin
@@ -6,6 +6,7 @@
 import vmem "core:mem/virtual"
 import "core:os"
 import "core:strings"
+import "core:strconv"
 import "core:unicode/utf8"
 
 when !ODIN_DEBUG { _ :: log } // Avoid unused import error for log
@@ -21,6 +22,7 @@
 	arena_allocator := vmem.arena_allocator(&arena)
 
 	input_buffer: [1024]byte
+
 	for {
 		defer free_all(context.temp_allocator)
 
@@ -56,6 +58,7 @@
 ////////////////////////////////////////////////////////////////////////////////
 
 TokenType :: enum {
+	Nil,
 	IdentifierVariable,
 	OperatorBinaryAssignment,
 	ConstantInteger,
@@ -282,6 +285,76 @@
 }
 
 ////////////////////////////////////////////////////////////////////////////////
+// Execution
+////////////////////////////////////////////////////////////////////////////////
+
+// Execute the tokens directly and return a map with variable name -> value.
+execute :: proc(tokens: [dynamic]Token, source_text: string, allocator: runtime.Allocator) -> map[string]int {
+	// Initialize interpreter state
+	var_values := make(map[string]int, allocator)
+
+	// Execution state
+	is_doing_assignment : bool
+	lefthand_identifier : Token
+	righthand_token : Token
+
+	for token in tokens {
+		switch token.type {
+		case .Nil:
+			continue
+		case .IdentifierVariable:
+			if is_doing_assignment {
+				righthand_token = token
+			} else {
+				lefthand_identifier = token
+			}
+		case .OperatorBinaryAssignment:
+			is_doing_assignment = true
+			// clear any righthand tokens
+			righthand_token = {}
+		case .ConstantInteger:
+			if is_doing_assignment {
+				righthand_token = token
+			}
+		}
+
+		// We have enough context to do an assignment operation!:
+		//     lefthand := righthand
+		if lefthand_identifier.type != .Nil && is_doing_assignment && righthand_token.type != .Nil {
+			lefthand_var_name := source_text[lefthand_identifier.start_byte:][:lefthand_identifier.length]
+			righthand_value : int
+			ok : bool
+			switch righthand_token.type {
+			case .IdentifierVariable:
+				righthand_var_name := source_text[righthand_token.start_byte:][:righthand_token.length]
+				if righthand_value, ok = var_values[righthand_var_name]; !ok {
+					unimplemented("Tried to read from variable before it had a value!")
+				}
+			case .ConstantInteger:
+				righthand_value_string := source_text[righthand_token.start_byte:][:righthand_token.length]
+				if righthand_value, ok = strconv.parse_int(righthand_value_string); !ok {
+					unimplemented(fmt.tprintfln("Can't parse constant integer %v", righthand_value_string))
+				}
+			case .Nil:
+				fallthrough
+			case .OperatorBinaryAssignment:
+				unimplemented(fmt.tprintfln("Can't assign a right hand value for token type %v", righthand_token.type))
+			}
+
+			// Do the actual assignment, overwriting anything already there
+			var_values[lefthand_var_name] = righthand_value
+
+			// Reset for next assignment operation
+			is_doing_assignment = false
+			lefthand_identifier = {}
+			righthand_token = {}
+		}
+	}
+
+	return var_values
+}
+
+////////////////////////////////////////////////////////////////////////////////
 // Tests
 ////////////////////////////////////////////////////////////////////////////////
 // odin test . -debug -define:ODIN_TEST_SHORT_LOGS=true -define:ODIN_TEST_THREADS=1
@@ -367,3 +440,15 @@
 	testing.expect_value(t, errors[1].line_number, 2)
 	testing.expect_value(t, errors[1].column_number, 14)
 }
+
+@(test)
+test_execute_009 :: proc(t: ^testing.T) {
+	source_text := "a := 1\nb := 3\nc := b\nd := c\n"
+	tokens, errors := tokenize(source_text, context.temp_allocator)
+	testing.expect_value(t, len(errors), 0)
+	var_values := execute(tokens, source_text, context.temp_allocator)
+	testing.expect_value(t, var_values["a"], 1)
+	testing.expect_value(t, var_values["b"], 3)
+	testing.expect_value(t, var_values["c"], 3)
+	testing.expect_value(t, var_values["d"], 3)
+}
