--- step-014/main.odin
+++ step-015/main.odin
@@ -73,8 +73,18 @@
 			}
 		}
 
+		ast_nodes, parsing_errors := parse(tokens, arena_allocator)
+
+		// Print out any errors
+		if len(parsing_errors) > 0 {
+			for error in parsing_errors {
+				fmt.eprint(generate_error_message(error, source_text))
+			}
+			continue
+		}
+
 		// Execute the tokens directly, for now
-		execute(tokens, source_text, &state, arena_allocator)
+		execute(ast_nodes, tokens, source_text, &state, arena_allocator)
 	}
 }
 
@@ -559,67 +569,43 @@
 	return state
 }
 
-// Execute the tokens directly and return a ProgramState struct, initializing state if nil
-execute :: proc(tokens: [dynamic]Token, source_text: string, state: ^InterpreterState, allocator: runtime.Allocator) {
-	// Execution state
-	is_doing_assignment : bool
-	lefthand_identifier : Token
-	righthand_token : Token
-
-	for token in tokens {
-		switch token.type {
+// Execute the AST
+execute :: proc(ast_nodes: [dynamic]AstNode, tokens: [dynamic]Token, source_text: string, state: ^InterpreterState, allocator: runtime.Allocator) {
+	// For now, just iterate a linear array of ast nodes
+	for ast_node in ast_nodes {
+		switch ast_node.type {
 		case .Nil:
 			continue
-		case .IdentifierVariable:
-			if is_doing_assignment {
-				righthand_token = token
-			} else {
-				lefthand_identifier = token
-			}
-		case .OperatorBinaryAssignment:
-			is_doing_assignment = true
-			// clear any righthand tokens
-			righthand_token = {}
-		case .ConstantInteger:
-			if is_doing_assignment {
-				righthand_token = token
-			}
-		}
-
-		// We have enough context to do an assignment operation!:
-		//     lefthand := righthand
-		if lefthand_identifier.type != .Nil && is_doing_assignment && righthand_token.type != .Nil {
-			lefthand_var_name := strings.clone(source_text[lefthand_identifier.start_byte:][:lefthand_identifier.length], allocator)
-			if !(lefthand_var_name in state.var_values) {
+		case .Assignment:
+			// destination := source
+			dest_token := tokens[ast_node.destToken]
+			dest_var_name := strings.clone(source_text[dest_token.line_start_byte:][:dest_token.length], allocator)
+			if !(dest_var_name in state.var_values) {
 				// First time we've seen this variable! Create it
-				append_elem(&state.var_names, lefthand_var_name)
+				append_elem(&state.var_names, dest_var_name)
 			}
-			righthand_value : int
+			source_token := tokens[ast_node.sourceToken]
+			source_value : int
 			ok : bool
-			switch righthand_token.type {
+			switch source_token.type {
 			case .IdentifierVariable:
-				righthand_var_name := source_text[righthand_token.start_byte:][:righthand_token.length]
-				if righthand_value, ok = state.var_values[righthand_var_name]; !ok {
+				righthand_var_name := source_text[source_token.start_byte:][:source_token.length]
+				if source_value, ok = state.var_values[righthand_var_name]; !ok {
 					unimplemented("Tried to read from variable before it had a value!")
 				}
 			case .ConstantInteger:
-				righthand_value_string := source_text[righthand_token.start_byte:][:righthand_token.length]
-				if righthand_value, ok = strconv.parse_int(righthand_value_string); !ok {
-					unimplemented(fmt.tprintfln("Can't parse constant integer %v", righthand_value_string))
+				source_value_string := source_text[source_token.start_byte:][:source_token.length]
+				if source_value, ok = strconv.parse_int(source_value_string); !ok {
+					unimplemented(fmt.tprintfln("Can't parse constant integer %v", source_value_string))
 				}
 			case .Nil:
 				fallthrough
 			case .OperatorBinaryAssignment:
-				unimplemented(fmt.tprintfln("Can't assign a right hand value for token type %v", righthand_token.type))
+				unimplemented(fmt.tprintfln("Can't assign a right hand value for token type %v", source_token.type))
 			}
 
 			// Do the actual assignment, overwriting anything already there
-			state.var_values[lefthand_var_name] = righthand_value
-
-			// Reset for next assignment operation
-			is_doing_assignment = false
-			lefthand_identifier = {}
-			righthand_token = {}
+			state.var_values[dest_var_name] = source_value
 		}
 	}
 }
@@ -730,8 +716,10 @@
 	source_text := "a := 1\nb := 3\nc := b\nd := c\n"
 	tokens, errors := tokenize(source_text, context.temp_allocator)
 	testing.expect_value(t, len(errors), 0)
+	ast_nodes, parsing_errors := parse(tokens, context.temp_allocator)
+	testing.expect_value(t, len(parsing_errors), 0)
 	state : InterpreterState = init_state(context.temp_allocator)
-	execute(tokens, source_text, &state, context.temp_allocator)
+	execute(ast_nodes, tokens, source_text, &state, context.temp_allocator)
 	testing.expect_value(t, state.var_names[0], "a")
 	testing.expect_value(t, state.var_values[state.var_names[0]], 1)
 	testing.expect_value(t, state.var_names[1], "b")
