--- step-014/main.odin
+++ step-015/main.odin
@@ -63,8 +63,18 @@
 			}
 		}
 
+		ast_nodes, parsing_errors := parse(tokens, arena_allocator)
+
+		// Print out any errors
+		if len(parsing_errors) > 0 {
+			for error in parsing_errors {
+				fmt.eprint(generate_error_message(error, input_string))
+			}
+			continue
+		}
+
 		// Execute the tokens directly, for now
-		execute(tokens, input_string, &state, arena_allocator)
+		execute(ast_nodes, tokens, input_string, &state, arena_allocator)
 	}
 }
 
@@ -549,41 +559,22 @@
 	return state
 }
 
-// Execute the tokens directly and return a ProgramState struct, initializing state if nil
-execute :: proc(tokens: [dynamic]Token, source_text: string, state: ^InterpreterState, allocator: runtime.Allocator) {
-	// Execution state
-	is_doing_assignment : bool
-	dest_identifier : Token
-	source_token : Token
-
-	for token in tokens {
-		switch token.type {
+// Execute the AST
+execute :: proc(ast_nodes: [dynamic]AstNode, tokens: [dynamic]Token, source_text: string, state: ^InterpreterState, allocator: runtime.Allocator) {
+	// For now, just iterate a linear array of ast nodes
+	for ast_node in ast_nodes {
+		switch ast_node.type {
 		case .Nil:
 			continue
-		case .IdentifierVariable:
-			if is_doing_assignment {
-				source_token = token
-			} else {
-				dest_identifier = token
-			}
-		case .OperatorBinaryAssignment:
-			is_doing_assignment = true
-			// clear any source tokens
-			source_token = {}
-		case .ConstantInteger:
-			if is_doing_assignment {
-				source_token = token
-			}
-		}
-
-		// We have enough context to do an assignment operation!:
-		// destination := source
-		if dest_identifier.type != .Nil && is_doing_assignment && source_token.type != .Nil {
-			dest_var_name := strings.clone(source_text[dest_identifier.start_byte:][:dest_identifier.length], allocator)
+		case .Assignment:
+			// destination := source
+			dest_token := tokens[ast_node.destToken]
+			dest_var_name := strings.clone(source_text[dest_token.line_start_byte:][:dest_token.length], allocator)
 			if !(dest_var_name in state.var_values) {
 				// First time we've seen this variable! Create it
 				append_elem(&state.var_names, dest_var_name)
 			}
+			source_token := tokens[ast_node.sourceToken]
 			source_value : int
 			ok : bool
 			switch source_token.type {
@@ -605,11 +596,6 @@
 
 			// Do the actual assignment, overwriting anything already there
 			state.var_values[dest_var_name] = source_value
-
-			// Reset for next assignment operation
-			is_doing_assignment = false
-			dest_identifier = {}
-			source_token = {}
 		}
 	}
 }
@@ -720,8 +706,10 @@
 	source_text := "a := 1\nb := 3\nc := b\nd := c\n"
 	tokens, errors := tokenize(source_text, context.temp_allocator)
 	testing.expect_value(t, len(errors), 0)
+	ast_nodes, parsing_errors := parse(tokens, context.temp_allocator)
+	testing.expect_value(t, len(parsing_errors), 0)
 	state : InterpreterState = init_state(context.temp_allocator)
-	execute(tokens, source_text, &state, context.temp_allocator)
+	execute(ast_nodes, tokens, source_text, &state, context.temp_allocator)
 	testing.expect_value(t, state.var_names[0], "a")
 	testing.expect_value(t, state.var_values[state.var_names[0]], 1)
 	testing.expect_value(t, state.var_names[1], "b")
