--- step-003/main.odin
+++ step-004/main.odin
@@ -1,6 +1,7 @@
 package lowkey
 
 import "core:fmt"
+import "core:log"
 
 main :: proc() {
 	fmt.println("Hello, world, from the Lowkey compiler!")
@@ -14,6 +15,9 @@
 	type: TokenType,
 	start_byte: int,
 	length: int,
+	line_number: int,
+	column_number: int,
+	token_index: int,
 }
 
 tokenize :: proc(source_text: string) -> [dynamic]Token {
@@ -22,6 +26,10 @@
 
 	output := make([dynamic]Token, context.temp_allocator)
 
+	line_number := 1
+	column_number := 0
+	token_index := 0
+
 	current_position := 0
 	next_position := 0
 	// A simple state machine to indicate if we are tokenizing a word or not as
@@ -32,6 +40,7 @@
 	for next_position < len(source_text) {
 		current_position = next_position
 		next_position += 1
+		column_number += 1
 
 		character := source_text[current_position]
 
@@ -40,9 +49,32 @@
 			// If we hit whitespace after an identifier, append token to output!
 			if tokenizing_word {
 				current_token.length = current_position - current_token.start_byte
+				when ODIN_DEBUG {
+					log.debugf(
+						"> Token %v: %v (%v:%v, byte %v) (%v)",
+						current_token.token_index,
+						source_text[current_token.start_byte:(current_token.start_byte + current_token.length)],
+						current_token.start_byte,
+						current_token.line_number,
+						current_token.column_number,
+						current_token.type
+					)
+				}
 				append(&output, current_token)
 				tokenizing_word = false
 			}
+
+			// A newline increments line number and resets column number
+			if character == '\n' {
+				line_number += 1
+				column_number = 0
+			}
+
+			when ODIN_DEBUG {
+				if character == '\n' {
+					log.debugf("--------------------------(newline)--------------------------")
+				}
+			}
 			continue
 		}
 
@@ -53,13 +85,16 @@
 				type = .IdentifierVariable,
 				start_byte = current_position,
 				// We will fill in length at end of parsing token
+				line_number = line_number,
+				column_number = column_number,
+				token_index = token_index,
 			}
+			token_index += 1
 			tokenizing_word = true
 		}
 	}
 
 	// Assume source text ends in whitespace, so final token gets appended
-
 	return output
 }
 
@@ -78,10 +113,44 @@
 
 @(test)
 test_tokenize :: proc(t: ^testing.T) {
-	source_text := "This   is my program\n"
+	source_text := "This   is my program\nLine two\n Line  three \n"
+	when ODIN_DEBUG {
+		log.debug("Testing tokenize() with source text:")
+		log.debug("---------------------------------------------------------")
+		log.debug(source_text)
+		log.debug("---------------------------------------------------------")
+	}
 	output := tokenize(source_text)
 	testing.expect_value(t, token_to_string(output[0], source_text), "This")
+	testing.expect_value(t, output[0].line_number, 1)
+	testing.expect_value(t, output[0].column_number, 1)
+	testing.expect_value(t, output[0].token_index, 0)
 	testing.expect_value(t, token_to_string(output[1], source_text), "is")
+	testing.expect_value(t, output[1].line_number, 1)
+	testing.expect_value(t, output[1].column_number, 8)
+	testing.expect_value(t, output[1].token_index, 1)
 	testing.expect_value(t, token_to_string(output[2], source_text), "my")
+	testing.expect_value(t, output[2].line_number, 1)
+	testing.expect_value(t, output[2].column_number, 11)
+	testing.expect_value(t, output[2].token_index, 2)
 	testing.expect_value(t, token_to_string(output[3], source_text), "program")
-}
\ No newline at end of file
+	testing.expect_value(t, output[3].line_number, 1)
+	testing.expect_value(t, output[3].column_number, 14)
+	testing.expect_value(t, output[3].token_index, 3)
+	testing.expect_value(t, token_to_string(output[4], source_text), "Line")
+	testing.expect_value(t, output[4].line_number, 2)
+	testing.expect_value(t, output[4].column_number, 1)
+	testing.expect_value(t, output[4].token_index, 4)
+	testing.expect_value(t, token_to_string(output[5], source_text), "two")
+	testing.expect_value(t, output[5].line_number, 2)
+	testing.expect_value(t, output[5].column_number, 6)
+	testing.expect_value(t, output[5].token_index, 5)
+	testing.expect_value(t, token_to_string(output[6], source_text), "Line")
+	testing.expect_value(t, output[6].line_number, 3)
+	testing.expect_value(t, output[6].column_number, 2)
+	testing.expect_value(t, output[6].token_index, 6)
+	testing.expect_value(t, token_to_string(output[7], source_text), "three")
+	testing.expect_value(t, output[7].line_number, 3)
+	testing.expect_value(t, output[7].column_number, 8)
+	testing.expect_value(t, output[7].token_index, 7)
+}
