--- step-003/main.odin
+++ step-004/main.odin
@@ -2,6 +2,9 @@
 
 import "base:runtime"
 import "core:fmt"
+import "core:log"
+
+when !ODIN_DEBUG { _ :: log } // Avoid unused import error for log
 
 main :: proc() {
 	fmt.println("Hello, world, from the Lowkey compiler!")
@@ -19,14 +22,27 @@
 	type: TokenType,
 	start_byte: int,
 	length: int,
+	line_number: int,
+	column_number: int,
+	token_index: int,
 }
 
 tokenize :: proc(source_text: string, allocator: runtime.Allocator) -> [dynamic]Token {
 	// TODO: Assert that source file length is < INT_MAX
 	// TODO: Assert that source file ends with whitespace
+	when ODIN_DEBUG {
+		log.debug("tokenize() source text:")
+		log.debug("---------------------------------------------------------")
+		log.debug(source_text)
+		log.debug("---------------------------------------------------------")
+	}
 
 	tokens := make([dynamic]Token, allocator)
 
+	line_number := 1
+	column_number := 0
+	token_index := 0
+
 	current_position := 0
 	next_position := 0
 	// A simple state machine to indicate if we are tokenizing a word or not as
@@ -37,6 +53,7 @@
 	for next_position < len(source_text) {
 		current_position = next_position
 		next_position += 1
+		column_number += 1
 
 		character := source_text[current_position]
 
@@ -45,9 +62,32 @@
 			if currently_tokenizing {
 				// If we hit whitespace after an identifier, append token to output!
 				current_token.length = current_position - current_token.start_byte
+				when ODIN_DEBUG {
+					log.debugf(
+						"> Token %v: %v (%v:%v, byte %v) (%v)",
+						current_token.token_index,
+						source_text[current_token.start_byte:(current_token.start_byte + current_token.length)],
+						current_token.line_number,
+						current_token.column_number,
+						current_token.start_byte,
+						current_token.type
+					)
+				}
 				append(&tokens, current_token)
 				currently_tokenizing = false
 			}
+
+			// A newline increments line number and resets column number
+			if character == '\n' {
+				line_number += 1
+				column_number = 0
+			}
+
+			when ODIN_DEBUG {
+				if character == '\n' {
+					log.debugf("--------------------------(newline)--------------------------")
+				}
+			}
 			continue
 		}
 
@@ -58,13 +98,16 @@
 				type = .IdentifierVariable,
 				start_byte = current_position,
 				// We will fill in length at end of parsing token
+				line_number = line_number,
+				column_number = column_number,
+				token_index = token_index,
 			}
+			token_index += 1
 			currently_tokenizing = true
 		}
 	}
 
 	// Assume source text ends in whitespace, so final token gets appended
-
 	return tokens
 }
 
@@ -85,10 +128,38 @@
 
 @(test)
 test_tokenize :: proc(t: ^testing.T) {
-	source_text := "This   is my program\n"
+	source_text := "This   is my program\nLine two\n Line  three \n"
 	tokens := tokenize(source_text, context.temp_allocator)
 	testing.expect_value(t, token_to_string(tokens[0], source_text), "This")
+	testing.expect_value(t, tokens[0].line_number, 1)
+	testing.expect_value(t, tokens[0].column_number, 1)
+	testing.expect_value(t, tokens[0].token_index, 0)
 	testing.expect_value(t, token_to_string(tokens[1], source_text), "is")
+	testing.expect_value(t, tokens[1].line_number, 1)
+	testing.expect_value(t, tokens[1].column_number, 8)
+	testing.expect_value(t, tokens[1].token_index, 1)
 	testing.expect_value(t, token_to_string(tokens[2], source_text), "my")
+	testing.expect_value(t, tokens[2].line_number, 1)
+	testing.expect_value(t, tokens[2].column_number, 11)
+	testing.expect_value(t, tokens[2].token_index, 2)
 	testing.expect_value(t, token_to_string(tokens[3], source_text), "program")
-}
\ No newline at end of file
+	testing.expect_value(t, tokens[3].line_number, 1)
+	testing.expect_value(t, tokens[3].column_number, 14)
+	testing.expect_value(t, tokens[3].token_index, 3)
+	testing.expect_value(t, token_to_string(tokens[4], source_text), "Line")
+	testing.expect_value(t, tokens[4].line_number, 2)
+	testing.expect_value(t, tokens[4].column_number, 1)
+	testing.expect_value(t, tokens[4].token_index, 4)
+	testing.expect_value(t, token_to_string(tokens[5], source_text), "two")
+	testing.expect_value(t, tokens[5].line_number, 2)
+	testing.expect_value(t, tokens[5].column_number, 6)
+	testing.expect_value(t, tokens[5].token_index, 5)
+	testing.expect_value(t, token_to_string(tokens[6], source_text), "Line")
+	testing.expect_value(t, tokens[6].line_number, 3)
+	testing.expect_value(t, tokens[6].column_number, 2)
+	testing.expect_value(t, tokens[6].token_index, 6)
+	testing.expect_value(t, token_to_string(tokens[7], source_text), "three")
+	testing.expect_value(t, tokens[7].line_number, 3)
+	testing.expect_value(t, tokens[7].column_number, 8)
+	testing.expect_value(t, tokens[7].token_index, 7)
+}
