--- step-006/main.odin
+++ step-007/main.odin
@@ -3,11 +3,46 @@
 import "base:runtime"
 import "core:fmt"
 import "core:log"
+import vmem "core:mem/virtual"
+import "core:os"
 
 when !ODIN_DEBUG { _ :: log } // Avoid unused import error for log
 
+////////////////////////////////////////////////////////////////////////////////
+// REPL
+////////////////////////////////////////////////////////////////////////////////
+
 main :: proc() {
-	fmt.println("Hello, world, from the Lowkey compiler!")
+	fmt.println("Welcome to the Lowkey compiler! Starting REPL mode:")
+
+	arena: vmem.Arena
+	arena_allocator := vmem.arena_allocator(&arena)
+
+	input_buffer: [1024]byte
+	for {
+		fmt.print("> ")
+		n, err := os.read(os.stdin, input_buffer[:])
+		if err != nil {
+			fmt.eprintln("Error reading:", err)
+			return
+		}
+
+		// Tokenize input
+		tokens, errors := tokenize(string(input_buffer[:n]), arena_allocator)
+
+		// Print out any errors
+		if len(errors) > 0 {
+			for error in errors {
+				fmt.eprintln("Error: ", error)
+			}
+			continue
+		}
+
+		// Print out output
+		for token in tokens {
+			fmt.println(token)
+		}
+	}
 }
 
 ////////////////////////////////////////////////////////////////////////////////
