--- step-002/main.odin
+++ step-003/main.odin
@@ -11,35 +11,69 @@
 // Tokenization
 ////////////////////////////////////////////////////////////////////////////////
 
-tokenize :: proc(source_text: string, allocator: runtime.Allocator) -> [dynamic]u8 {
-	output := make([dynamic]u8, allocator)
-	next_position := 0
+TokenType :: enum {
+	IdentifierVariable,
+}
+
+Token :: struct {
+	type: TokenType,
+	start_byte: int,
+	length: int,
+}
 
+tokenize :: proc(source_text: string, allocator: runtime.Allocator) -> [dynamic]Token {
+	// TODO: Assert that source file length is < INT_MAX
+	// TODO: Assert that source file ends with whitespace
+
+	tokens := make([dynamic]Token, allocator)
+
+	current_position := 0
+	next_position := 0
 	// A simple state machine to indicate if we are tokenizing a word or not as
 	// we loop through characters.
 	tokenizing_word := false
-
+	current_token : Token
 	// Loop through each byte
 	for next_position < len(source_text) {
-		current_position := next_position
+		current_position = next_position
 		next_position += 1
 
 		character := source_text[current_position]
 
 		// Skip all whitespace
-		if character == ' ' || character == '\t' || character == '\n' {
-			tokenizing_word = false
+		if is_whitespace(character) {
+			// If we hit whitespace after an identifier, append token to output!
+			if tokenizing_word {
+				current_token.length = current_position - current_token.start_byte
+				append(&tokens, current_token)
+				tokenizing_word = false
+			}
 			continue
 		}
 
-		// Anything other than whitespace is a word! (for now)
+		// Anything other than whitespace is an identifier! (for now)
 		if !tokenizing_word {
 			// Append byte position of new word to output
-			append(&output, u8(current_position))
+			current_token = Token {
+				type = .IdentifierVariable,
+				start_byte = current_position,
+				// We will fill in length at end of parsing token
+			}
 			tokenizing_word = true
 		}
 	}
-	return output
+
+	// Assume source text ends in whitespace, so final token gets appended
+
+	return tokens
+}
+
+token_to_string :: proc(token: Token, source_text: string) -> string {
+	return source_text[token.start_byte:(token.start_byte + token.length)]
+}
+
+is_whitespace :: proc(character: u8) -> bool {
+	return character == ' ' || character == '\t' || character == '\n'
 }
 
 ////////////////////////////////////////////////////////////////////////////////
@@ -52,9 +86,9 @@
 @(test)
 test_tokenize :: proc(t: ^testing.T) {
 	source_text := "This   is my program\n"
-	output := tokenize(source_text, context.temp_allocator)
-	testing.expect_value(t, output[0], 0)
-	testing.expect_value(t, output[1], 7)
-	testing.expect_value(t, output[2], 10)
-	testing.expect_value(t, output[3], 13)
+	tokens := tokenize(source_text, context.temp_allocator)
+	testing.expect_value(t, token_to_string(tokens[0], source_text), "This")
+	testing.expect_value(t, token_to_string(tokens[1], source_text), "is")
+	testing.expect_value(t, token_to_string(tokens[2], source_text), "my")
+	testing.expect_value(t, token_to_string(tokens[3], source_text), "program")
 }
\ No newline at end of file
