--- step-012/main.odin
+++ step-013/main.odin
@@ -89,13 +89,19 @@
 	ConstantInteger,
 }
 
+TokenFlag :: enum {
+	IsStatementEnd, // If set, then this token is the last token in a statement
+}
+
 Token :: struct {
 	type: TokenType,
 	start_byte: int,
 	length: int,
 	line_number: int,
 	column_number: int,
+	line_start_byte: int,
 	token_index: int,
+	flags: bit_set[TokenFlag],
 }
 
 // The states for the tokenization state machine
@@ -148,7 +154,13 @@
 
 				if tokenization_state == .CurrentlyTokenizing {
 					// Allow for a comment to come right after a token. E.g. a := 5// Hi
+					current_token.flags += {.IsStatementEnd}
 					complete_and_append_token(&tokens, &current_token, current_position, source_text)
+				} else {
+					// Mark the last previously-completed token on this line as a statement end. E.g. a := 5 // Hi
+					if len(tokens) > 0 && tokens[len(tokens) - 1].line_number == line_number {
+						tokens[len(tokens) - 1].flags += {.IsStatementEnd}
+					}
 				}
 
 				tokenization_state = .PreviousSlash
@@ -182,6 +194,7 @@
 
 		// Skip all whitespace
 		if is_whitespace(character) {
+			finish_token := false
 			// If we hit whitespace after an identifier, append token to output!
 			if tokenization_state == .CurrentlyComment {
 				when ODIN_DEBUG {
@@ -189,19 +202,27 @@
 				}
 				// Don't do anything until newline is hit
 			} else if tokenization_state == .CurrentlyTokenizing {
-				complete_and_append_token(&tokens, &current_token, current_position, source_text)
+				finish_token = true
 				tokenization_state = .Idle
 			} else if tokenization_state == .CurrentlyError {
 				// Reset now that we are going to next token
 				tokenization_state = .Idle
 			}
 
-			// A newline increments line number and resets column number
+			// A newline increments line number, resets column number, and marks token as statement end
 			if character == '\n' {
 				line_number += 1
 				column_number = 0
 				line_start_byte = current_position + 1
 				tokenization_state = .Idle // Reset single-line comment
+				// For the last token in a line, add a 'statement end' flag
+				if finish_token {
+					current_token.flags += {.IsStatementEnd}
+				}
+			}
+
+			if finish_token {
+				complete_and_append_token(&tokens, &current_token, current_position, source_text)
 			}
 
 			when ODIN_DEBUG {
@@ -222,6 +243,7 @@
 				// We will fill in length at end of parsing token
 				line_number = line_number,
 				column_number = column_number,
+				line_start_byte = line_start_byte,
 				token_index = token_index,
 			}
 
@@ -295,9 +317,15 @@
 // Errors
 ////////////////////////////////////////////////////////////////////////////////
 
+
+// ErrorType and Error are shared between the tokenizer and the parser
 ErrorType :: enum {
 	TokenizationInvalidNumber,
 	TokenizationStraySlash,
+	ParsingStrayNumber,
+	ParsingStrayIdentifier,
+	ParsingStrayAssignment,
+	ParsingMultipleAssignment,
 }
 
 Error :: struct {
@@ -326,6 +354,14 @@
 		strings.write_string(&builder, "Tokenization: Invalid Number")
 	case .TokenizationStraySlash:
 		strings.write_string(&builder, "Tokenization: Stray Slash")
+	case .ParsingStrayNumber:
+		strings.write_string(&builder, "Parsing: Stray Number")
+	case .ParsingStrayIdentifier:
+		strings.write_string(&builder, "Parsing: Stray Identifier")
+	case .ParsingStrayAssignment:
+		strings.write_string(&builder, "Parsing: Stray Assignment")
+	case .ParsingMultipleAssignment:
+		strings.write_string(&builder, "Parsing: Multiple Assignments on same line")
 	}
 	strings.write_string(&builder, " (")
 	strings.write_int(&builder, error.line_number)
@@ -387,8 +423,12 @@
 	NumberConstant
 }
 
-parse :: proc(tokens: [dynamic]Token, source_text: string, allocator: runtime.Allocator) -> ([dynamic]AstNode) {
+parse :: proc(tokens: [dynamic]Token, source_text: string, allocator: runtime.Allocator) -> ([dynamic]AstNode, [dynamic]Error) {
 	nodes := make([dynamic]AstNode, allocator)
+	errors := make([dynamic]Error, allocator)
+
+	// Simple parsing state
+	last_assignment_line : int
 
 	curr_token := 0
 	curr_statement := 0
@@ -399,27 +439,78 @@
 		case .Nil:
 			unimplemented()
 		case .IdentifierVariable:
+			if TokenFlag.IsStatementEnd in token.flags {
+				append(&errors, Error {
+					type = .ParsingStrayIdentifier,
+					start_byte = token.start_byte,
+					line_start_byte = token.line_start_byte,
+					line_number = token.line_number,
+					column_number = token.column_number,
+				})
+				curr_token += 1
+				continue
+			}
+
 			if curr_token + 1 >= total_tokens {
 				// Not enough tokens left to finish assignment
 				continue
 			}
 			next_token := tokens[curr_token + 1]
+
 			if next_token.type == .OperatorBinaryAssignment {
+				if next_token.line_number == last_assignment_line {
+					append(&errors, Error {
+						type = .ParsingMultipleAssignment,
+						start_byte = token.start_byte,
+						line_start_byte = token.line_start_byte,
+						line_number = token.line_number,
+						column_number = token.column_number,
+					})
+
+					// Advance tokens forward until a "end statement" token is found
+					for curr_token < total_tokens {
+						if tokens[curr_token].flags == {.IsStatementEnd} {
+							curr_token += 1
+							break
+						}
+						curr_token += 1
+					}
+					continue
+				}
+
 				node, tokens_eaten := parse_statement_assignment(tokens, curr_token)
 				append_elem(&nodes, node)
 				curr_token += tokens_eaten
 				curr_statement += 1
+				// Make sure multiple assignments are never done on the same line
+				last_assignment_line = next_token.line_number
 			} else {
 				unimplemented()
 			}
 		case .OperatorBinaryAssignment:
-			unimplemented()
+			// We can never start with an assignment token
+			append(&errors, Error {
+				type = .ParsingStrayAssignment,
+				start_byte = token.start_byte,
+				line_start_byte = token.line_start_byte,
+				line_number = token.line_number,
+				column_number = token.column_number,
+			})
+			curr_token += 1
 		case .ConstantInteger:
-			unimplemented()
+			// Can never start with a constant integer
+			append(&errors, Error {
+				type = .ParsingStrayNumber,
+				start_byte = token.start_byte,
+				line_start_byte = token.line_start_byte,
+				line_number = token.line_number,
+				column_number = token.column_number,
+			})
+			curr_token += 1
 		}
 	}
 
-	return nodes
+	return nodes, errors
 }
 
 // Parse an assignment statement and return an AstNode and the number of tokens it consumed
@@ -667,7 +758,8 @@
 	source_text := "a := 1\nb := 3\nc := b\nd := c\n"
 	tokens, tokenization_errors := tokenize(source_text, context.temp_allocator)
 	testing.expect_value(t, len(tokenization_errors), 0)
-	statements := parse(tokens, source_text, context.temp_allocator)
+	statements, parsing_errors := parse(tokens, source_text, context.temp_allocator)
+	testing.expect_value(t, len(parsing_errors), 0)
 	testing.expect_value(t, len(statements), 4)
 	testing.expect_value(t, statements[0].type, AstNodeType.Assignment)
 	testing.expect_value(t, statements[0].destType, ValueType.Variable)
@@ -726,3 +818,29 @@
 	testing.expect_value(t, errors[0].line_number, 9)
 	testing.expect_value(t, errors[0].column_number, 1)
 }
+
+@(test)
+test_parse_013 :: proc(t: ^testing.T) {
+	source_text :=
+`a := 5 // This is valid
+1 // Can't have a constant number by itself
+:= // Can't have an assignment by itself
+a // Can't have a variable by itself
+a := b c := d // Can't have multiple assignments on the same line
+e := f := g // Can't have an assignment that starts with nothing
+`
+	tokens, tokenization_errors := tokenize(source_text, context.temp_allocator)
+	testing.expect_value(t, len(tokenization_errors), 0)
+	statements, parsing_errors := parse(tokens, source_text, context.temp_allocator)
+	testing.expect_value(t, len(statements), 3)
+	testing.expect_value(t, statements[0].type, AstNodeType.Assignment)
+	testing.expect_value(t, statements[0].destType, ValueType.Variable)
+	testing.expect_value(t, token_index_to_string(statements[0].destToken, tokens, source_text), "a")
+	testing.expect_value(t, statements[0].sourceType, ValueType.NumberConstant)
+	testing.expect_value(t, token_index_to_string(statements[0].sourceToken, tokens, source_text), "5")
+	testing.expect_value(t, parsing_errors[0].type, ErrorType.ParsingStrayNumber)
+	testing.expect_value(t, parsing_errors[1].type, ErrorType.ParsingStrayAssignment)
+	testing.expect_value(t, parsing_errors[2].type, ErrorType.ParsingStrayIdentifier)
+	testing.expect_value(t, parsing_errors[3].type, ErrorType.ParsingMultipleAssignment)
+	testing.expect_value(t, parsing_errors[4].type, ErrorType.ParsingStrayAssignment)
+}
