--- step-009/main.odin
+++ step-010/main.odin
@@ -11,6 +11,9 @@
 
 when !ODIN_DEBUG { _ :: log } // Avoid unused import error for log
 
+KB :: 1024
+MB :: KB * 1024
+
 ////////////////////////////////////////////////////////////////////////////////
 // REPL
 ////////////////////////////////////////////////////////////////////////////////
@@ -18,8 +21,13 @@
 main :: proc() {
 	fmt.println("Welcome to the Lowkey compiler! Starting REPL mode:")
 
+	// program-wide allocations
 	arena: vmem.Arena
 	arena_allocator := vmem.arena_allocator(&arena)
+	state := init_state(arena_allocator)
+
+	// Make a source text string builder, and init to large capacity so it doesn't reallocate
+	source_text_builder := strings.builder_make_len_cap(1 * MB, 50 * MB, arena_allocator)
 
 	input_buffer: [1024]byte
 
@@ -33,7 +41,19 @@
 			return
 		}
 
-		source_text := string(input_buffer[:n])
+		input_string := string(input_buffer[:n])
+
+		// Handle any commands (not intended to be part of the source text)
+		if input_string == "state\n" {
+			print_interpreter_state(&state)
+			continue
+		}
+
+		// Append input string to source text
+		strings.write_string(&source_text_builder, input_string)
+
+		// Get a string view/slice into the source text
+		source_text := strings.to_string(source_text_builder)
 
 		// Tokenize input
 		tokens, errors := tokenize(source_text, arena_allocator)
@@ -46,10 +66,15 @@
 			continue
 		}
 
-		// Print out output
-		for token in tokens {
-			fmt.println(token)
+		when ODIN_DEBUG {
+			// Print out tokens
+			for token in tokens {
+				fmt.println(token)
+			}
 		}
+
+		// Execute the tokens directly, for now
+		execute(tokens, source_text, &state, arena_allocator)
 	}
 }
 
@@ -283,11 +308,23 @@
 // Execution
 ////////////////////////////////////////////////////////////////////////////////
 
-// Execute the tokens directly and return a map with variable name -> value.
-execute :: proc(tokens: [dynamic]Token, source_text: string, allocator: runtime.Allocator) -> map[string]int {
-	// Initialize interpreter state
-	var_values := make(map[string]int, allocator)
+InterpreterState :: struct {
+	// Keep track of variable names, in order of creation
+	var_names: [dynamic]string,
+	// Keep track of variable values
+	var_values: map[string]int,
+}
+
+init_state :: proc(allocator: runtime.Allocator) -> InterpreterState {
+	state := InterpreterState {
+		var_names = make([dynamic]string, allocator),
+		var_values = make(map[string]int, allocator),
+	}
+	return state
+}
 
+// Execute the tokens directly and return a ProgramState struct, initializing state if nil
+execute :: proc(tokens: [dynamic]Token, source_text: string, state: ^InterpreterState, allocator: runtime.Allocator) {
 	// Execution state
 	is_doing_assignment : bool
 	dest_identifier : Token
@@ -316,13 +353,17 @@
 		// We have enough context to do an assignment operation!:
 		// destination := source
 		if dest_identifier.type != .Nil && is_doing_assignment && source_token.type != .Nil {
-			dest_var_name := source_text[dest_identifier.start_byte:][:dest_identifier.length]
+			dest_var_name := strings.clone(source_text[dest_identifier.start_byte:][:dest_identifier.length], allocator)
+			if !(dest_var_name in state.var_values) {
+				// First time we've seen this variable! Create it
+				append_elem(&state.var_names, dest_var_name)
+			}
 			source_value : int
 			ok : bool
 			switch source_token.type {
 			case .IdentifierVariable:
 				source_var_name := source_text[source_token.start_byte:][:source_token.length]
-				if source_value, ok = var_values[source_var_name]; !ok {
+				if source_value, ok = state.var_values[source_var_name]; !ok {
 					unimplemented(fmt.tprintfln("Tried to read from variable '%v' before it had a value!", source_var_name))
 				}
 			case .ConstantInteger:
@@ -337,7 +378,7 @@
 			}
 
 			// Do the actual assignment, overwriting anything already there
-			var_values[dest_var_name] = source_value
+			state.var_values[dest_var_name] = source_value
 
 			// Reset for next assignment operation
 			is_doing_assignment = false
@@ -345,10 +386,23 @@
 			source_token = {}
 		}
 	}
+}
+
+print_interpreter_state :: proc (state: ^InterpreterState) {
+	builder := strings.builder_make(context.temp_allocator)
+	strings.write_string(&builder, "Interpreter State:\n")
+
+	for variable in state.var_names {
+		strings.write_string(&builder, variable)
+		strings.write_string(&builder, " = ")
+		strings.write_int(&builder, state.var_values[variable])
+		strings.write_rune(&builder, '\n')
+	}
 
-	return var_values
+	fmt.print(strings.to_string(builder))
 }
 
+
 ////////////////////////////////////////////////////////////////////////////////
 // Tests
 ////////////////////////////////////////////////////////////////////////////////
@@ -441,9 +495,14 @@
 	source_text := "a := 1\nb := 3\nc := b\nd := c\n"
 	tokens, errors := tokenize(source_text, context.temp_allocator)
 	testing.expect_value(t, len(errors), 0)
-	var_values := execute(tokens, source_text, context.temp_allocator)
-	testing.expect_value(t, var_values["a"], 1)
-	testing.expect_value(t, var_values["b"], 3)
-	testing.expect_value(t, var_values["c"], 3)
-	testing.expect_value(t, var_values["d"], 3)
+	state : InterpreterState = init_state(context.temp_allocator)
+	execute(tokens, source_text, &state, context.temp_allocator)
+	testing.expect_value(t, state.var_names[0], "a")
+	testing.expect_value(t, state.var_values[state.var_names[0]], 1)
+	testing.expect_value(t, state.var_names[1], "b")
+	testing.expect_value(t, state.var_values[state.var_names[1]], 3)
+	testing.expect_value(t, state.var_names[2], "c")
+	testing.expect_value(t, state.var_values[state.var_names[2]], 3)
+	testing.expect_value(t, state.var_names[3], "d")
+	testing.expect_value(t, state.var_values[state.var_names[3]], 3)
 }
